### DISCUSSION TITLE
yarn build fails with new server components as data is fetched from nextjs API routes

### DISCUSSION THREAD

**Discussion by chris-hinds** [vercel/next.js#43311] [App Router]
Currently experimenting with NextJS 13 and its server components with component-level data fetching.

I am fetching data from the next API routes and the build fails as these are obviously not available at build time.

Can we not turn off the pre-rendering during build time? I just want server-side rendering not for next to try and build the pages during build.

**Comment by icyJoseph** [comment]
Well, in a sense you are doing something the docs call unnecessary and inefficient, https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props#getserversideprops-or-api-routes:

> It can be tempting to reach for an [API Route](https://nextjs.org/docs/api-routes/introduction) when you want to fetch data from the server, then call that API route from getServerSideProps. This is an unnecessary and inefficient approach, as it will cause an extra request to be made due to both getServerSideProps and API Routes running on the server.

And it is not only a property of Next 13, but rather has always been there since Next 9. It is not only that the API routes are not available, but that you are strictly requiring an extra round-trip, when you could just get the data at the component already.

## Alternative

Assuming you've come around the absolute URLs requirement to fetch from the server, then you could tell Nextjs to make this page on-deman, via https://beta.nextjs.org/docs/api-reference/segment-config#dynamic, **although it seems to have a bug**, after all this is a beta stage release. The idea is that it'd not do any rendering with this page during build time, only at runtime.

**Reply by chris-hinds** [reply]
Thanks, @icyJoseph I actually did try to force the dynamic pages as per those docs however this doesn't seem to be working.

I generally prefer to have most logic for data fetch on the API routes as I sometimes need/want to format the data and I do not want that within my FE code (I know the data fetching code is only on the server but it's just a preference of mine). Plus I am fetching this data from an external DB. I also do not want to interface directly with the DB in this FE code

**Reply by icyJoseph** [reply]
> I actually did try to force the dynamic pages as per those docs however this doesn't seem to be working

Yeah I think it is still forming... aka buggy 😅 Remember that this is a beta release, you are not expected to take this all the way to production, even less so in critical apps.

> I generally prefer to have most logic for data fetch on the API routes as I sometimes need/want to format the data and I do not want that within my FE code (I know the data fetching code is only on the server but it's just a preference of mine). Plus I am fetching this data from an external DB. I also do not want to interface directly with the DB in this FE code

Well that's a matter of how you look at it isn't it? You could parametrize your data fetching so that you could transparently call it within an API route, or from a Server Component, or even middleware. At the end of the day you import `fetchMyData`, calculate the inputs in the given context, and get your data.
